### DISCUSSION TITLE
How should I implement Data Validation in ListBox?

### DISCUSSION THREAD

**Discussion by kyuranger** [AvaloniaUI/Avalonia#15329] [Q&A] [answered]
```
public class TestClass:ObservableValidator
{
string _SerialPortName = "";
[SerialPortName]
public string SerialPortName {
    get => _SerialPortName;
    set { 
        if (value !=_SerialPortName) { 
            _SerialPortName = value;
            OnPropertyChanged();                    
            ValidateProperty(SerialPortName);
        }
    }
}

public sealed class SerialPortNameAttribute : ValidationAttribute
{
    protected override ValidationResult IsValid(object value, ValidationContext validationContext)
    {
        if (value is string _string&&!string.IsNullOrEmpty(_string)
        {
            return ValidationResult.Success;
        }
        return new("Invalid SerialPort");
    }
}
}
```
When the program is running, although the SerialPortNameAttribute is running normally, no error message appears. It looks like this is because there are no DataValidationErrors in the controltemplate of listbox. I'm not quite sure how to solve this problem.

![image](https://github.com/AvaloniaUI/Avalonia/assets/30720036/1e79d824-2d0f-4d0d-ad42-efeaa653444b)
PS: It seems that if the ValidationAttribute detects an error, it will automatically add the error class to the listbox. I think maybe I can take advantage of this. I can add a textblock and then decide whether to display it by judging whether there is this class, but I don't know how to bind it.

**Comment by maxkatz6** [comment]
ListBox doesn't have error styles built-in, as it wasn't expected for single selection control to have a validation error.
The easiest approach would be to implement your own simple validation in the view model and display it with your own bind text on the UI.

**Reply by kyuranger** [reply, marked as answer]
It seems that this is all we can do for the time being. Thank you very much for your answer.
